scanpy/tools/_dendrogram.py

```diff
@@ -103,49 +112,70 @@ def dendrogram(
+                )
+            if not isinstance(adata.obs[group].dtype, CategoricalDtype):
+                raise ValueError(
+                    "groupby has to be a categorical observation. "
+                    f"Given value: {group}, Column type: {adata.obs[group].dtype}"
+                )
+
+        if var_names is None:
+            rep_df = pd.DataFrame(
+                _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)
             )
-        if not isinstance(adata.obs[group].dtype, CategoricalDtype):
-            raise ValueError(
-                "groupby has to be a categorical observation. "
-                f"Given value: {group}, Column type: {adata.obs[group].dtype}"
-            )
-
-    if var_names is None:
-        rep_df = pd.DataFrame(
-            _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)
-        )
-        categorical = adata.obs[groupby[0]]
-        if len(groupby) > 1:
-            for group in groupby[1:]:
-                # create new category by merging the given groupby categories
-                categorical = (
-                    categorical.astype(str) + "_" + adata.obs[group].astype(str)
-                ).astype("category")
-        categorical.name = "_".join(groupby)
-
-        rep_df.set_index(categorical, inplace=True)
-        categories = rep_df.index.categories
+            categorical = adata.obs[groupby[0]]
+            if len(groupby) > 1:
+                for group in groupby[1:]:
+                    # create new category by merging the given groupby categories
+                    categorical = (
+                        categorical.astype(str) + "_" + adata.obs[group].astype(str)
+                    ).astype("category")
+            categorical.name = "_".join(groupby)
+
+            rep_df.set_index(categorical, inplace=True)
+            categories = rep_df.index.categories
+        else:
+            from scanpy.plotting._anndata import _prepare_dataframe
+
+            categories, rep_df = _prepare_dataframe(adata, var_names, groupby, use_raw)
+
+        # aggregate values within categories using 'mean'
+        rep_df = rep_df.groupby(level=0).mean()
     else:
-        gene_names = adata.raw.var_names if use_raw else adata.var_names
-        from ..plotting._anndata import _prepare_dataframe
-
-        categories, rep_df = _prepare_dataframe(adata, gene_names, groupby, use_raw)
+        if var_names is None:
+            rep_df = pd.DataFrame(
+                _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)
+            )
 
-    # aggregate values within categories using 'mean'
-    mean_df = rep_df.groupby(level=0).mean()
+        else:
+            rep_df = pd.DataFrame(
+                _choose_representation(
+                    adata[:, var_names], use_rep=use_rep, n_pcs=n_pcs
+                )
+            )
+        categories = rep_df.axes[axis]
 
     import scipy.cluster.hierarchy as sch
     from scipy.spatial import distance
 
```

Comment:
nit and unrelated to this PR, but I think this should go first in the function.  The function can't run without them and the `groupby` call can be expensive.  It happens before this, so no point wasting people's time if they somehow don't have these packages installed (I understand that they are dependencies so that shouldn't happen, but still).

Reply:
> Not sure if there is done all over scanpy, but wouldn't wrapping this in a try-catch block with a useful error message be good?
> I understand that they are dependencies so that shouldn't happen, but still

Why still? Of course all hard dependencies (like scipy) are expected to be installed properly. No need to handle this in any special way.

The inline import is probably for performance reasons.

Reply:
You're right, there really isn't a "still" here.  That makes sense